forc-util/src/restricted.rs

```diff
@@ -130,3 +144,42 @@ fn test_invalid_char() {
         std::result::Result::Ok(())
     ));
 }
+
+#[test]
```

Comment:
Perhaps something like below is a bit cleaner?

```rust
#[test]
fn test_is_valid_project_name_format() {
    let assert_valid = |name: &str| {
        is_valid_project_name_format(name).expect("this should pass");
    };

    let assert_invalid = |name: &str, expected_error: &str| {
        assert_eq!(
            is_valid_project_name_format(name).map_err(|e| e.to_string()),
            Err(expected_error.into())
        );
    };

    // Test valid project names
    assert_valid("mock_project_name");
    assert_valid("mock_project_name123");
    assert_valid("mock_project_name-123-_");

    // Test invalid project names
    let error_message = "the project name `{}` cannot be used as a project name.\n\
        project name can be a combination of letters, numbers, hyphen, and underscores, \
        and must start with a letter.";

    assert_invalid("1mock_project", &format!(error_message, "1mock_project"));
    assert_invalid("mock_.project", &format!(error_message, "mock_.project"));
    assert_invalid("mock_/project", &format!(error_message, "mock_/project"));
}
```

Reply:
@JoshuaBatty Hi sir, I think `&format!(error_message, "1mock_project")` would issue an error "format argument must be a string literal".

I wrapped the format logic into a function
```rust
    let format_error_message = |name: &str| -> String {
        let formatted_err_str = format!(
            "the project name `{}` cannot be used as a project name.\n\
        project name can be a combination of letters, numbers, hyphen, and underscores, \
        and must start with a letter.",
            name
        );
        formatted_err_str
    };
```